add adot bidder adapter

mlequain opened the pull request:
<!--
Thank you for your pull request. Please make sure this PR is scoped to one change, and that any added or changed code includes tests with greater than 80% code coverage. See https://github.com/prebid/Prebid.js/blob/master/CONTRIBUTING.md#testing-prebidjs for documentation on testing Prebid.js.
-->

## Type of change
<!-- Remove items that don't apply and/or select an item by changing [ ] to [x] -->
- [X] New bidder adapter  <!--  IMPORTANT: if checking here, also submit your bidder params documentation here https://github.com/prebid/prebid.github.io/tree/master/dev-docs/bidders --> 

## Description of change
<!-- Describe the change proposed in this pull request -->
Add adot bidder adapter

<!-- For new bidder adapters, please provide the following -->
- test parameters for validating bids
```
{
  bidder: 'adot',
  params: {
    placementId: 'adot_prebidjs_integration'
  }
}
```

Be sure to test the integration with your adserver using the [Hello World](/integrationExamples/gpt/hello_world.html) sample page.

- contact email of the adapter’s maintainer: maxime.lequain@we-are-adot.com
- [X] official adapter submission

For any changes that affect user-facing APIs or example code documented on http://prebid.org, please provide:

- Documentation PR: https://github.com/prebid/prebid.github.io/pull/1846

jsnellbaker commented on modules/adotBidAdapter.js:
```diff
@@ -0,0 +1,636 @@
+  try {
+    return JSON.parse(data);
+  } catch (err) {
+    console.error(err);
```
Please use the `src/utils` `logError` function instead of a direct console call.

jsnellbaker commented on modules/adotBidAdapter.js:
```diff
@@ -0,0 +1,636 @@
+  getUserSyncs(syncOptions, serverResponses) {},
+  onTimeout(timeoutData) {},
+  onBidWon(ad) {},
+  onSetTargeting(ad) {}
```
If you don't plan to use these functions as part of your adapter, you can remove them.  They don't have to be defined.

jsnellbaker commented on modules/adotBidAdapter.js:
```diff
@@ -0,0 +1,636 @@
+
+    return bidResponse.seatbid
+      .filter(seatBid => isObject(seatBid) && isArray(seatBid.bid))
+      .flatMap(seatBid => seatBidsToAds(seatBid, bidResponse, serverRequest));
```
`flatMap` is also not supported in IE11.  Can you please use an alternative?

jsnellbaker commented on modules/adotBidAdapter.js:
```diff
@@ -0,0 +1,636 @@
+      .then(vastXml => done(null, vastXml))
+      .catch(err => done(err, null));
+
+    return;
```
I'm afraid we don't support Promises in Prebid.js.  Can you please rewrite this logic without promises?  

Additionally, (if I'm understanding the larger workflow) would it be possible to do this overall logic for your outstream player without it requiring an extra network request just to download the VAST XML?  If there's some way the raw VAST could be returned as part of the bid response from your endpoint, that would be ideal.

jsnellbaker commented on modules/adotBidAdapter.js:
```diff
@@ -0,0 +1,636 @@
+          .then(() => player.startAd())
+          .catch(err => console.error(err));
+      });
+    });
```
Similar to the earlier comment, I'm afraid we don't support Promises in Prebid.js.  Can you please rewrite this code in a different way?

jsnellbaker commented on modules/adotBidAdapter.js:
```diff
@@ -0,0 +1,636 @@
+function validateVideo(video) {
+  if (!isObject(video)) return false;
+  if (!isString(video.context)) return false;
+  if (!SUPPORTED_VIDEO_CONTEXTS.includes(video.context)) return false;
```
The `includes` here isn't supported by IE11.  Is there an alternative you can use?

jsnellbaker commented on modules/adotBidAdapter.js:
```diff
@@ -0,0 +1,636 @@
+
+  if (isInstream(adUnit.mediaTypes.video)) {
+    if (!video.instreamContext) return false;
+    if (!SUPPORTED_INSTREAM_CONTEXTS.includes(video.instreamContext)) return false;
```
Please see earlier note.

jsnellbaker commented on modules/adotBidAdapter.js:
```diff
@@ -0,0 +1,636 @@
+  if (!isObject(bid.ext)) return false;
+  if (!isObject(bid.ext.adot)) return false;
+  if (!isString(bid.ext.adot.media_type)) return false;
+  if (!BID_SUPPORTED_MEDIA_TYPES.includes(bid.ext.adot.media_type)) return false;
```
Please see earlier note.

jsnellbaker commented on modules/adotBidAdapter.js:
```diff
@@ -0,0 +1,636 @@
+  if (!bid.adm && !bid.nurl) return false;
+  if (bid.adm) {
+    if (!isString(bid.adm)) return false;
+    if (!bid.adm.includes(AUCTION_PRICE)) return false;
```
Please see earlier note.

jsnellbaker commented on modules/adotBidAdapter.js:
```diff
@@ -0,0 +1,636 @@
+  }
+  if (bid.nurl) {
+    if (!isString(bid.nurl)) return false;
+    if (!bid.nurl.includes(AUCTION_PRICE)) return false;
```
Please see earlier note.

jsnellbaker commented on modules/adotBidAdapter.js:
```diff
@@ -0,0 +1,636 @@
+    const allowedBidderCodes = [this.code];
+
+    return isObject(adUnit) &&
+      allowedBidderCodes.includes(adUnit.bidder) &&
```
Please see earlier note.

jsnellbaker commented on test/spec/modules/adotBidAdapter_spec.js:
```diff
@@ -0,0 +1,3179 @@
+import { expect } from 'chai';
+import { executeRenderer } from 'src/Renderer';
+import * as utils from 'src/utils';
+import { spec } from 'modules/adotBidAdapter';
```
Please include the `.js` extension on these imports as well (eg. `src/Renderer.js`).

mlequain commented on modules/adotBidAdapter.js:
```diff
@@ -0,0 +1,636 @@
+function validateMediaSizes(mediaSize) {
+  return isArray(mediaSize) &&
+    (mediaSize.length === 2) &&
+    mediaSize.every(size => (isNumber(size) && size >= 0));
```
The issue with `isArrayOfNums` is that it does not check if the value is superior to 0.

mlequain commented on modules/adotBidAdapter.js:
```diff
@@ -0,0 +1,636 @@
+  return Object
+    .entries(groups)
+    .map(([id, values]) => ({id, key, values}));
+}
```
The function in the `src/utils.js` does not format the values the same way unfortunately. Do you want me to rename the method to avoid confusion ?

mlequain commented on modules/adotBidAdapter.js:
```diff
@@ -0,0 +1,636 @@
+}
+
+function isBanner(mediaTypes) {
+  return isObject(mediaTypes) && isObject(mediaTypes.banner);
```
In most of our uses cases we also want to check for a specific type or look inside arrays which does not seem to be covered by the deepAccess method unless i'm mistaken

mlequain commented:
@jsnellbaker I've updated the code to address the issues you mentioned in your previous comment.

Let me know if there is anything that I missed or if you want me to change anything else.

mlequain commented:
Hi @jsnellbaker 

Sorry for the late reply.

I've updated the code to remove the calls to `find` and use the polyfill instead, let me know if everything is OK.

Below you'll find the adUnits you can use to test the video (outstream & instream) as well as the native.

## Instream adUnit

```javascript
const adUnit = {
    code: 'test-div',
    mediaTypes: {
        video: {
            context: 'instream',
            playerSize: [[300, 250]]
        }
    },
    bids: [{
        bidder: 'adot',
        params: {
            video: {
                mimes: ['video/mp4'],
                minDuration: 5,
                maxDuration: 35,
                protocols: [2, 3],
                instreamContext: 'pre-roll'
            },
            placementId: 'adot_prebidjs_integration'
        }
    }]
}
```

## Outstream adUnit

```javascript
const adUnit = {
    code: 'test-div',
    mediaTypes: {
        video: {
            context: 'outstream',
            playerSize: [[300, 250]]
        }
    },
    bids: [{
        bidder: 'adot',
        params: {
            video: {
                mimes: ['video/mp4'],
                minDuration: 5,
                maxDuration: 35,
                protocols: [2, 3]
            },
            placementId: 'adot_prebidjs_integration'
        }
    }]
}
```

## Native adUnit

```javascript
const adUnit = {
    code: 'test-div',
    mediaTypes: {
        native: {
            native: {
                image: {
                    required: false,
                    sizes: [100, 50]
                },
                title: {
                    required: false,
                    len: 140
                },
                sponsoredBy: {
                    required: false
                },
                clickUrl: {
                    required: false
                },
                body: {
                    required: false
                },
                icon: {
                    required: false,
                    sizes: [50, 50]
                }
            }
        }
    },
    bids: [{
        bidder: 'adot',
        params: {
    	    placementId: 'adot_prebidjs_integration'
        }
    }]
}
```

mlequain commented:
Hi @jsnellbaker 

I have updated the code to make sure we either return a `vastUrl` or a `vastXml` for all the video bids, the instream video should now be returning a valid prebid ad.

The outstream video is working fine on my end.

I managed to bid on the native ad using your bid request, let me know if it's working on your end as well.

jsnellbaker commented:
@mlequain 

Thanks for the follow-up.  I retested the video instream and saw it was working.  

However, I tried the native test again, but I'm still not seeing a bid return.  

I noticed the sample params had an extra `native` object under the `mediaTypes`.  When I removed that, I saw the native asset data was properly added to the `imp[0].native.request.assets` (where before it wasn't).  But this didn't appear to help return a bid.

I have copied the updated request payload below, can you take a look and advise?
```
{"id":"1d746d3476ea1e","imp":[{"id":"2ba0032f0a9c5a_0","native":{"request":"{\"assets\":[{\"id\":3,\"required\":false,\"img\":{\"type\":3,\"wmin\":100,\"hmin\":50}},{\"id\":1,\"required\":false,\"title\":{\"len\":140}},{\"id\":4,\"required\":false,\"data\":{\"type\":1}},{\"id\":5,\"required\":false,\"data\":{\"type\":2}},{\"id\":2,\"required\":false,\"img\":{\"type\":1,\"wmin\":50,\"hmin\":50}}]}"},"pmp":{"deals":[{"id":"adot_prebidjs_integration"}]}}],"site":{"page":"http://test.localhost:9999/integrationExamples/gpt/demo_native.html?pbjs_debug=true","domain":"test.localhost:9999","name":"test.localhost:9999"},"device":{"ua":"Mozilla/5.0 (iPhone; CPU iPhone OS 13_2_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0.3 Mobile/15E148 Safari/604.1","language":"en-US"},"user":null,"regs":null,"at":1,"ext":{"adot":{"adapter_version":"v1.0.0"},"should_use_gzip":true}}
```

mlequain commented:
@jsnellbaker 

I have updated the adapter documentation and my previous comment to fix the mistake in the native snippet.

I have also updated some targeting parameters on my end and I'm seeing bids for the native ad using your bid request.

Let me know if everything is working on your end.

Thanks a lot

jsnellbaker commented:
@mlequain Thanks for the additional update, I am getting a native bid back now.  LGTM

mlequain commented:
@jsnellbaker Thanks a lot for your help !
